File: Array/238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        

        int[] out = new int[nums.length];
        
        Arrays.fill(out, 1);

        int leftProduct = 1;

        for(int i=1;i<nums.length;i++) {
            leftProduct *= nums[i-1];
            out[i] = leftProduct;
        }

        int rightProduct = 1;

        for(int i=nums.length-2;i>=0;i--) {
            rightProduct *= nums[i+1];
            out[i] *= rightProduct;
        }

        return out;


    }
}

File: Array/238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        out = [1]*len(nums)
        left = 1
        for i in range(1,len(nums)):
            left *= nums[i-1]
            out[i] = left
        right = 1
        for i in range(len(nums)-2, -1, -1):
            right *= nums[i+1]
            out[i]*= right
        return out
        
